### ISSUE TITLE
Create a unifying Reader trait for the Raw/System/User levels

### ISSUE THREAD

**Issue by zslayton** [amzn/ion-rust#345]
The observations below are @almann's thoughts from PR #344.

-----

**TL;DR:** These are some API ergonomics that the addition of the `SystemReader` has brought to light for me.  I don't think this should block merging, but serve as some observations for future work around the API.  I will cut an issue from this missive.

----

I noticed our APIs around this are slightly non-orthogonal.   For example:

```rust
// in trait RawReader
fn next(&mut self) -> IonResult<Option<RawStreamItem>>;

// in SystemReader
pub fn next(&mut self) -> IonResult<Option<SystemStreamItem>> { ... }

// in Reader
pub fn next(&mut self) -> IonResult<Option<(IonType, bool)>> { ... }
```

These enum definitions also are related semantically but not in code:

```rust
pub enum RawStreamItem {
    VersionMarker(u8, u8),
    Value(IonType, bool),
}

pub enum SystemStreamItem {
    VersionMarker(u8, u8),
    SymbolTableData(IonType, bool),
    Value(IonType, bool),
}

// not actually modeled as enum, but as a tuple in the actual API
// added here because it is effectively modeling a part of the above
pub enum StreamItem {
    Value(IonType, bool)
}
```

So my question is this--should we eliminate `Option` and align these enums?  E.g. (spelling nonwithstanding),

```rust
pub enum RawStreamItem {
    VersionMarker(u8, u8),
    Value(IonType, bool),
    Nothing,
}

pub enum SystemStreamItem {
    VersionMarker(u8, u8),
    SymbolTableData(IonType, bool),
    Value(IonType, bool),
    Nothing,
}

pub enum UserStreamItem {
    Value(IonType, bool),
    Nothing,
}

pub trait StreamItem {
    // either marker trait or nice ways to normalize the variants above...
    ...
}
```

It is debatable to me if we want to model `Nothing` as a value that is either `End`/`Before` or something--I think that state difference _could_ be useful.

This could make it so that we can define what is now the `RawReader` trait as just `Reader` with an associated type `type Item: StreamItem;` or some such and make `next(&mut self) -> Self::Item`.  This way the higher level implementations (`Reader`/`SystemReader`) and the underlying implementations (`RawBinaryReader`/`RawTextReader`) can all implement the trait.  I think we'd probably need an associated type for the symbol token as well.

Technically the generalization of the trait doesn't required my flattening above (nor the bounds on the associated type), but I think it might make the API a bit cleaner and less wrapped in nested enums.

Another alternative would be to wrap the enums in common, which keeps the definitions associated with one another, but I think that is uglier in net.

All of this is a long winded way of saying we should reconsider how we factor the stream item enum and factor the `RawReader` trait such that it can be easily used by all implementations.  The benefits are pretty clear on the latter (the enum flattening is optional for a more generic trait), we can actually now write generic code against the system reader and the regular reader and the raw readers (which allows someone else to do decoration or other interesting things).

_Originally posted by @almann in https://github.com/amzn/ion-rust/pull/344#discussion_r779139401_
